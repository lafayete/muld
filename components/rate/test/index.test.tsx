import * as React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup, fireEvent } from '@testing-library/react';
import 'jest-styled-components';
import Rate from '..';

function mockGetBoundingClientRect(items: HTMLCollection) {
    for (let i = 0; i < items.length; i++) {
        items[i].getBoundingClientRect = () =>
            ({
                left: i * 20,
                width: 20,
            } as any);
    }
}
const NewRate = (props: any) => {
    const { initValue, jestFn, ...others } = props;
    const [value, setValue] = React.useState<number>(initValue);
    const onChange = (v: number) => {
        jestFn(v);
        setValue(v);
    };
    return <Rate value={value} onChange={onChange} {...others}></Rate>;
};

describe('Rate', () => {
    afterEach(cleanup);

    it('onChange event', () => {
        const jestFn = jest.fn();
        const { container } = render(<NewRate initValue={1} jestFn={jestFn} />);
        const fourthIcon = container.querySelectorAll('.mul-rate__icon')[3];
        fireEvent.click(fourthIcon);
        expect(jestFn).toHaveBeenCalledWith(4);
        expect(jestFn).toHaveBeenCalledTimes(1);
    });

    it('allow half', () => {
        const jestFn = jest.fn();
        const { container } = render(<NewRate initValue={1} jestFn={jestFn} allowHalf={true} />);
        const fourthIcon = container.querySelectorAll('.mul-rate__icon--half')[3];
        fireEvent.click(fourthIcon);
        expect(jestFn).toHaveBeenCalledWith(3.5);
        expect(jestFn).toHaveBeenCalledTimes(1);
    });

    it('disabled', () => {
        const jestFn = jest.fn();
        const { container } = render(<NewRate initValue={1} jestFn={jestFn} disabled={true} />);
        mockGetBoundingClientRect(container.querySelector('.mul-rate')!.children);
        const icons = container.querySelectorAll('.mul-rate__icon');
        const fourthIcon = icons[3];
        fireEvent.click(fourthIcon);
        fireEvent.touchStart(fourthIcon, { touches: [{ clientX: 0, clientY: 0 }] });
        fireEvent.touchMove(fourthIcon, { touches: [{ clientX: 50, clientY: 10 }] });
        fireEvent.touchEnd(fourthIcon);
        expect(jestFn).toHaveBeenCalledTimes(0);
    });

    it('touch move to select item', () => {
        const jestFn = jest.fn();
        const { container } = render(<NewRate initValue={0} jestFn={jestFn} />);
        mockGetBoundingClientRect(container.querySelector('.mul-rate')!.children);
        const icons = container.querySelectorAll('.mul-rate__icon');
        const firstIcon = icons[0];
        fireEvent.touchStart(firstIcon, { touches: [{ clientX: 1, clientY: 0 }] });
        fireEvent.touchMove(firstIcon, { touches: [{ clientX: 21, clientY: 0 }] });
        fireEvent.touchMove(firstIcon, { touches: [{ clientX: 41, clientY: 0 }] });
        fireEvent.touchMove(firstIcon, { touches: [{ clientX: 61, clientY: 0 }] });
        fireEvent.touchMove(firstIcon, { touches: [{ clientX: 80, clientY: 0 }] });
        fireEvent.touchEnd(firstIcon);
        expect(jestFn).toHaveBeenCalledTimes(4);
        expect(jestFn).toHaveBeenNthCalledWith(1, 1);
        expect(jestFn).toHaveBeenNthCalledWith(2, 2);
        expect(jestFn).toHaveBeenNthCalledWith(3, 3);
        expect(jestFn).toHaveBeenNthCalledWith(4, 4);
    });

    it('touch move to select half item', () => {
        const jestFn = jest.fn();
        const { container } = render(<NewRate initValue={0} jestFn={jestFn} allowHalf={true} />);
        mockGetBoundingClientRect(container.querySelector('.mul-rate')!.children);
        const icons = container.querySelectorAll('.mul-rate__icon');
        const firstIcon = icons[0];
        fireEvent.touchStart(firstIcon, { touches: [{ clientX: 1, clientY: 0 }] });
        fireEvent.touchMove(firstIcon, { touches: [{ clientX: 11, clientY: 0 }] });
        fireEvent.touchMove(firstIcon, { touches: [{ clientX: 21, clientY: 0 }] });
        fireEvent.touchMove(firstIcon, { touches: [{ clientX: 31, clientY: 0 }] });
        fireEvent.touchMove(firstIcon, { touches: [{ clientX: 41, clientY: 0 }] });
        fireEvent.touchMove(firstIcon, { touches: [{ clientX: 50, clientY: 0 }] });
        fireEvent.touchEnd(firstIcon);
        expect(jestFn).toHaveBeenCalledTimes(5);
        expect(jestFn).toHaveBeenNthCalledWith(1, 0.5);
        expect(jestFn).toHaveBeenNthCalledWith(2, 1);
        expect(jestFn).toHaveBeenNthCalledWith(3, 1.5);
        expect(jestFn).toHaveBeenNthCalledWith(4, 2);
        expect(jestFn).toHaveBeenNthCalledWith(5, 2.5);
    });

    it('gutter prop', () => {
        const jestFn = jest.fn();
        const { container } = render(<NewRate initValue={0} jestFn={jestFn} gutter={10} />);
        expect(container).toMatchSnapshot();
    });

    it('size prop', () => {
        const jestFn = jest.fn();
        const { container } = render(<NewRate initValue={0} jestFn={jestFn} size="2rem" />);
        expect(container).toMatchSnapshot();
    });

    it('untouchable', () => {
        const jestFn = jest.fn();
        const { container } = render(<NewRate initValue={0} jestFn={jestFn} touchable={false} />);
        mockGetBoundingClientRect(container.querySelector('.mul-rate')!.children);
        const icons = container.querySelectorAll('.mul-rate__icon');
        const firstIcon = icons[0];
        fireEvent.touchStart(firstIcon, { touches: [{ clientX: 1, clientY: 0 }] });
        fireEvent.touchMove(firstIcon, { touches: [{ clientX: 21, clientY: 0 }] });
        fireEvent.touchMove(firstIcon, { touches: [{ clientX: 41, clientY: 0 }] });
        fireEvent.touchMove(firstIcon, { touches: [{ clientX: 61, clientY: 0 }] });
        fireEvent.touchMove(firstIcon, { touches: [{ clientX: 80, clientY: 0 }] });
        fireEvent.touchEnd(firstIcon);
        expect(jestFn).toHaveBeenCalledTimes(0);
    });
});
