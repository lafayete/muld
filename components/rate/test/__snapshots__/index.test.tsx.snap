// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rate gutter prop 1`] = `
.c1.mul-icon {
  position: relative;
  display: inline-block;
  font-family: 'muld-icon' !important;
  font-size: inherit;
  text-rendering: auto;
  -webkit-font-smoothing: antialiased;
}

.c1.mul-icon::before {
  display: inline-block;
}

.c1.mul-icon .mul-icon__image {
  width: 1em;
  height: 1em;
  object-fit: contain;
}

.c1.mul-icon-star-o::before {
  content: '\\F0C3';
}

.c0.mul-rate {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0.mul-rate .mul-rate__item {
  position: relative;
}

.c0.mul-rate .mul-rate__item:not(:last-child) {
  padding-right: 4px;
}

.c0.mul-rate .mul-rate__icon {
  display: block;
  width: 1em;
  color: #c8c9cc;
  font-size: 20px;
}

.c0.mul-rate .mul-rate__icon--half {
  position: absolute;
  top: 0;
  left: 0;
  width: 0.5em;
  overflow: hidden;
}

.c0.mul-rate .mul-rate__icon--full {
  color: #fa220a;
}

.c0.mul-rate .mul-rate__icon--disabled {
  color: #c8c9cc;
}

.c0.mul-rate.mul-rate--disabled {
  cursor: not-allowed;
}

.c0.mul-rate.mul-rate--readonly {
  cursor: default;
}

<div>
  <div
    class="c0 mul-rate"
  >
    <div
      class="mul-rate__item"
      style="padding-right: 10px;"
    >
      <i
        class="c1 mul-icon mul-icon-star-o mul-rate__icon"
        content="F0C3"
        name="star-o"
        style="color: rgb(200, 201, 204); font-size: 20px;"
      />
    </div>
    <div
      class="mul-rate__item"
      style="padding-right: 10px;"
    >
      <i
        class="c1 mul-icon mul-icon-star-o mul-rate__icon"
        content="F0C3"
        name="star-o"
        style="color: rgb(200, 201, 204); font-size: 20px;"
      />
    </div>
    <div
      class="mul-rate__item"
      style="padding-right: 10px;"
    >
      <i
        class="c1 mul-icon mul-icon-star-o mul-rate__icon"
        content="F0C3"
        name="star-o"
        style="color: rgb(200, 201, 204); font-size: 20px;"
      />
    </div>
    <div
      class="mul-rate__item"
      style="padding-right: 10px;"
    >
      <i
        class="c1 mul-icon mul-icon-star-o mul-rate__icon"
        content="F0C3"
        name="star-o"
        style="color: rgb(200, 201, 204); font-size: 20px;"
      />
    </div>
    <div
      class="mul-rate__item"
    >
      <i
        class="c1 mul-icon mul-icon-star-o mul-rate__icon"
        content="F0C3"
        name="star-o"
        style="color: rgb(200, 201, 204); font-size: 20px;"
      />
    </div>
  </div>
</div>
`;

exports[`Rate size prop 1`] = `
.c1.mul-icon {
  position: relative;
  display: inline-block;
  font-family: 'muld-icon' !important;
  font-size: inherit;
  text-rendering: auto;
  -webkit-font-smoothing: antialiased;
}

.c1.mul-icon::before {
  display: inline-block;
}

.c1.mul-icon .mul-icon__image {
  width: 1em;
  height: 1em;
  object-fit: contain;
}

.c1.mul-icon-star-o::before {
  content: '\\F0C3';
}

.c0.mul-rate {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0.mul-rate .mul-rate__item {
  position: relative;
}

.c0.mul-rate .mul-rate__item:not(:last-child) {
  padding-right: 4px;
}

.c0.mul-rate .mul-rate__icon {
  display: block;
  width: 1em;
  color: #c8c9cc;
  font-size: 20px;
}

.c0.mul-rate .mul-rate__icon--half {
  position: absolute;
  top: 0;
  left: 0;
  width: 0.5em;
  overflow: hidden;
}

.c0.mul-rate .mul-rate__icon--full {
  color: #fa220a;
}

.c0.mul-rate .mul-rate__icon--disabled {
  color: #c8c9cc;
}

.c0.mul-rate.mul-rate--disabled {
  cursor: not-allowed;
}

.c0.mul-rate.mul-rate--readonly {
  cursor: default;
}

<div>
  <div
    class="c0 mul-rate"
  >
    <div
      class="mul-rate__item"
      style="padding-right: 4px;"
    >
      <i
        class="c1 mul-icon mul-icon-star-o mul-rate__icon"
        content="F0C3"
        name="star-o"
        style="color: rgb(200, 201, 204); font-size: 2rem;"
      />
    </div>
    <div
      class="mul-rate__item"
      style="padding-right: 4px;"
    >
      <i
        class="c1 mul-icon mul-icon-star-o mul-rate__icon"
        content="F0C3"
        name="star-o"
        style="color: rgb(200, 201, 204); font-size: 2rem;"
      />
    </div>
    <div
      class="mul-rate__item"
      style="padding-right: 4px;"
    >
      <i
        class="c1 mul-icon mul-icon-star-o mul-rate__icon"
        content="F0C3"
        name="star-o"
        style="color: rgb(200, 201, 204); font-size: 2rem;"
      />
    </div>
    <div
      class="mul-rate__item"
      style="padding-right: 4px;"
    >
      <i
        class="c1 mul-icon mul-icon-star-o mul-rate__icon"
        content="F0C3"
        name="star-o"
        style="color: rgb(200, 201, 204); font-size: 2rem;"
      />
    </div>
    <div
      class="mul-rate__item"
    >
      <i
        class="c1 mul-icon mul-icon-star-o mul-rate__icon"
        content="F0C3"
        name="star-o"
        style="color: rgb(200, 201, 204); font-size: 2rem;"
      />
    </div>
  </div>
</div>
`;
